doc/src/modules/polys/numberfields.rst

```diff
@@ -0,0 +1,289 @@
+    $\alpha = f(\beta)$ for some $f(x) \in \mathbb{Q}[x]$.
+
+At present only a subset of the tasks enumerated in the lists above is yet
+supported in SymPy, and if you are interested in expanding support, you are
+encouraged to contribute!
+
+At time of writing, the existing solutions to the above problems are found
+in the following places:
+
+=================================  ======================================
+Task                               Implementation
+=================================  ======================================
+(1) integral basis                 :py:func:`~.round_two`
+(2) prime decomposition            :py:func:`~.prime_decomp`
+(3) $\mathfrak{p}$-adic valuation  :py:func:`~.prime_valuation`
+(10) find min poly                 :py:func:`~.minpoly`
+(11) isolating intervals           :py:func:`~.isolate`
+(12) primitive element             :py:func:`~.primitive_element`
+(13) field isomorphism             :py:func:`~.field_isomorphism`
+(14) field membership              :py:func:`~.to_number_field`
+=================================  ======================================
+
+
+Internals
+=========
+
+Algebraic number fields
+-----------------------
+
+Algebraic number fields are represented in SymPy by the
+:py:class:`~.AlgebraicField` class, which is a part of the
+polynomial domains system.
+
+
+
+Representing algebraic numbers
+------------------------------
+
+There are several different ways to represent algebraic numbers, and different
+forms may be preferable for different computational tasks.
+See [Cohen93]_ Section 4.2.
+
+
+As number field elements
+````````````````````````
+
+In SymPy, there is a distinction between number and expression classes defined
+in the :py:mod:`sympy.core.numbers` module on the one hand, and domains and
+domain elements defined in the :py:mod:`~sympy.polys` module on the other.
+
+When it comes to algebraic numbers, the :py:mod:`sympy.core.numbers` module
+offers the :py:class:`~.AlgebraicNumber` class, while the
+:py:mod:`~sympy.polys` module offers the
+:py:class:`~sympy.polys.polyclasses.ANP` class. This is the type of domain
+elements belonging to the :py:class:`~.AlgebraicField` domain.
+
+
+As points in the complex plane
+``````````````````````````````
+
+The minimal polynomial for an algebraic number determines it only up to
+conjugacy; in other words, the number and all its conjugates are algebraically
+indistinguishable.
+
+In order to select a particular conjugate, we can identify
+it with a point in the complex plane. This means identifying two real
+intervals, one on the real axis and one on the imaginary axis, such that
+exactly one root of the given irreducible polynomial over $\mathbb{Q}$ lies
+within the complex region so defined.
+
+Support for this in SymPy is currently limited. The :py:func:`~.isolate`
+function is currently applicable only to algebraic numbers lying on the
+real line.
```

Comment:
How exactly is this different from RootOf?
```python
In [8]: rs = [RootOf(x**5 - x + 1, n) for n in range(5)]

In [9]: [r.n(4) for r in rs]
Out[9]: [-1.167, -0.1812 - 1.084⋅ⅈ, -0.1812 + 1.084⋅ⅈ, 0.7649 - 0.3525⋅ⅈ, 0.7649 + 0.3525⋅ⅈ]

In [10]: rs[0]
Out[10]: 
       ⎛ 5           ⎞
CRootOf⎝x  - x + 1, 0⎠

In [11]: rs[0].is_real
Out[11]: True

In [12]: rs[1].is_real
Out[12]: False

In [13]: rs[1].conjugate() == rs[2]
Out[13]: True
```

Reply:
Interesting. TBH, on the code that I didn't write (anything in `numberfields.(minpoly|numbers|isomorphism)` I was doing my best to comment, without knowing the full story. I'll look into this some more.

Reply:
The code underlying RootOf is in `sympy/polys/rootisolation.py` e.g. `dup_isolate_all_roots`:
```python
In [1]: from sympy.polys.rootisolation import dup_isolate_all_roots

In [2]: dup_isolate_all_roots([ZZ(1), ZZ(0), ZZ(2)], ZZ)
Out[2]: 
([],
 [(((mpq(0,1), mpq(-4,1)), (mpq(4,1), mpq(0,1))), 1),
  (((mpq(0,1), mpq(0,1)), (mpq(4,1), mpq(4,1))), 1)])
```
This shows that `x**2 + 2` has no real roots and two complex roots of multiplicity 1 which are bounded by `(0, 4) + (-4, 0)*I` and `(0, 4) + (0, 4)*I`

Reply:
I have looked more closely at this funny `isolate()` function under `numberfields`, and here's what I find:

It is not used at all in SymPy (outside of the one unit test devoted to it), but it is
included in `from sympy import *`.

My best guess is that it exists to support something involving NumPy, SciPy, or other.
This is suggested by the use of `lambdify()` in its implementation. (Docstring for `lambdify` talks
about making SymPy objects usable by NumPy etc.)


Q: How does it compare to `RootOf` / `dup_isolate_all_roots`?

A: It _uses_ such things internally, and its purpose is to accept an algebraic number (not a poly),
_find_ a poly for that number, and then do some isolation.


Q: Why the restriction to _real_ algebraic numbers, since SymPy clearly does support isolating boxes in `CC`?

A: Not clear. Probably pertains to the intended external application, involving NumPy or whatever it was.


Having now reviewed this, I'd like to (a) improve the docstring, and (b) move the function. Please see my
latest comment on this PR.